src/Cooked/Skeleton.hs

```diff
@@ -388,35 +395,38 @@ type RedeemerConstrs redeemer =
     Typeable redeemer
   )
 
-data TxSkelRedeemer where
-  TxSkelNoRedeemer :: TxSkelRedeemer
-  TxSkelRedeemerForScript :: (RedeemerConstrs redeemer) => redeemer -> TxSkelRedeemer
-  -- | The first argument is a reference to the output where the reference
-  -- script is stored.
-  TxSkelRedeemerForReferenceScript :: (RedeemerConstrs redeemer) => Api.TxOutRef -> redeemer -> TxSkelRedeemer
-
-deriving instance (Show TxSkelRedeemer)
-
-txSkelTypedRedeemer :: (Api.FromData (Script.RedeemerType a)) => TxSkelRedeemer -> Maybe (Script.RedeemerType a)
-txSkelTypedRedeemer (TxSkelRedeemerForScript redeemer) = Api.fromData . Api.toData $ redeemer
-txSkelTypedRedeemer (TxSkelRedeemerForReferenceScript _ redeemer) = Api.fromData . Api.toData $ redeemer
-txSkelTypedRedeemer _ = Nothing
+data Redeemer where
+  NoRedeemer :: Redeemer
+  SomeRedeemer :: (RedeemerConstrs redeemer) => redeemer -> Redeemer
 
-txSkelReferenceScript :: TxSkelRedeemer -> Maybe Api.TxOutRef
-txSkelReferenceScript (TxSkelRedeemerForReferenceScript refScript _) = Just refScript
-txSkelReferenceScript _ = Nothing
+deriving instance (Show Redeemer)
 
-instance Eq TxSkelRedeemer where
-  TxSkelNoRedeemer == TxSkelNoRedeemer = True
-  (TxSkelRedeemerForScript r1) == (TxSkelRedeemerForScript r2) =
+instance Eq Redeemer where
+  NoRedeemer == NoRedeemer = True
+  (SomeRedeemer r1) == (SomeRedeemer r2) =
     case typeOf r1 `eqTypeRep` typeOf r2 of
       Just HRefl -> r1 PlutusTx.== r2
       Nothing -> False
-  (TxSkelRedeemerForReferenceScript o1 r1) == (TxSkelRedeemerForReferenceScript o2 r2) =
-    TxSkelRedeemerForScript r1 == TxSkelRedeemerForScript r2
-      && o1 == o2
   _ == _ = False
 
+data TxSkelRedeemer = TxSkelRedeemer
+  { txSkelRedeemer :: Redeemer,
+    txSkelReferenceScript :: Maybe Api.TxOutRef
+  }
+  deriving (Show, Eq)
+
+pattern TxSkelRedeemerForScript a = TxSkelRedeemer (SomeRedeemer a) Nothing
+
+pattern TxSkelNoRedeemer = TxSkelRedeemer NoRedeemer Nothing
+
+pattern TxSkelRedeemerForReferenceScript outRef a = TxSkelRedeemer (SomeRedeemer a) (Just outRef)
+
+pattern TxSkelNoRedeemerForReferenceScript outRef = TxSkelRedeemer NoRedeemer (Just outRef)
```

Comment:
I don't think these synonyms bring much value in this very simple case. I would find it as readable to pattern match on `TxSkelRedeemer` directly, especially since it is pretty straightforward.

I have the impression those are mostly for the benefit of the user, to expose more verbose constructors. I don't think the user will often pattern match on `TxSkelRedeemer` (which patterns would be good for [^1]) but they will mostly use these to construct values in skeleton. In this latter case, simple smart constructors would do[^2].

[^1]: Even in this case, for this usage, this would seem to me as overkill since pattern matching without synonyms is already very clear. I think this would be an overkill usage of pattern synonyms which, IMO, are relevant is more complex situations such as in [Haskell sequences](https://hackage.haskell.org/package/containers-0.6.6/docs/src/Data.Sequence.Internal.html)  where it really adds value with additional patterns for double-linked lists (that are not available from the type constructors) in a context where users will pattern match a lot. Here they are just aliases. Besides, we already heavily rely on many language extensions and I think we should take it down a notch and abstain when we can.

[^2]: I am not opposed to smart constructors but we should ask ourselves if they would really bring enough quality of life compared to bare usage of `TxSkelRedeemer`. I am all for smart constructors when they bring automatic data conversion or beyond-type safety. However, when they are just aliases with barely less characters than without them, maybe less is more.